File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/E-Расстояние_Дамерау-Левенштейна/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(NM)
    """
    a: str = sys.stdin.readline().rstrip()
    b: str = sys.stdin.readline().rstrip()
    n, m = len(a), len(b)
    dp: list[list[list[int]]] = [[[float('inf')] * 2 for _ in range(m + 1)] for _ in range(n + 1)]
    for i in range(n + 1):
        dp[i][0][0] = i
        dp[i][0][1] = i
    for j in range(m + 1):
        dp[0][j][0] = j
        dp[0][j][1] = j
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            dp[i][j][0] = min(
                dp[i - 1][j - 1][0] + int(a[i - 1] != b[j - 1]),
                dp[i - 1][j][0] + 1,
                dp[i][j - 1][0] + 1,
            )
            if i >= 2 and j >= 2 and a[i - 1] == b[j - 2] and a[i - 2] == b[j - 1]:
                dp[i][j][1] = dp[i - 2][j - 2][0] + 1
            else: dp[i][j][1] = float('inf')
            dp[i][j][0] = min(dp[i][j][0], dp[i][j][1])
    sys.stdout.write('{}\n'.format(dp[n][m][0]))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/E-Расстояние_Дамерау-Левенштейна/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <climits>
#include <algorithm>

void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(NM)
    */
    std::string a, b;
    std::cin >> a >> b;
    int n = a.size(), m = b.size();
    std::vector<std::vector<std::vector<int>>> dp(n + 1, std::vector<std::vector<int>>(m + 1, std::vector<int>(2, INT_MAX)));
    for (int i = 0; i <= n; ++i) {
        dp[i][0][0] = i;
        dp[i][0][1] = i;
    }
    for (int j = 0; j <= m; ++j) {
        dp[0][j][0] = j;
        dp[0][j][1] = j;
    }

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            dp[i][j][0] = std::min({
                dp[i-1][j-1][0] + (a[i-1] != b[j-1]),
                dp[i-1][j][0] + 1,
                dp[i][j-1][0] + 1,
            });

            if (i >= 2 && j >= 2 && a[i-1] == b[j-2] && a[i-2] == b[j-1]) {
                dp[i][j][1] = dp[i-2][j-2][0] + 1;
            } else {
                dp[i][j][1] = INT_MAX;
            }
            dp[i][j][0] = std::min(dp[i][j][0], dp[i][j][1]);
        }
    }
    std::printf("%d\n", dp[n][m][0]);
}

int main() {
    solution();
}